# 1.1 Amazon SageMaker 기본 Training 실습

## 학습 작업의 실행 노트북 개요
이 노트북은 주요하게 아래의 작업을 수행 합니다.

- 로컬 머신에서 로컬 모드 및 클라우드 모드로 모델 훈련
    - 환결 설정은 아래 링크를 참조 하세요.
        - [로컬 머신에서 Visual Studio Code 를 사용하여 SageMaker 훈련 코드 작성 하기](https://github.com/gonsoomoon-ml/Self-Study-On-SageMaker/blob/main/Environment/README-Local-VS-Code.md)
- SageMaker Notebook 에서 로컬 모드 및 클라우드 모드로 모델 훈련


이 노트북은 SageMaker에서 Training Job을 통해서 모델 학습합니다.상세한 사항은 개발자 가이드를 참조 하세요. -->  [모델 학습](https://sagemaker.readthedocs.io/en/stable/overview.html#prepare-a-training-script)




### 작업 실행 시 필요 라이브러리 import

In [65]:
import boto3
import sagemaker

### Bucket 정의

In [66]:
use_default_bucket = True

if use_default_bucket:
    s3_bucket = sagemaker.session.Session().default_bucket()
else:
    s3_bucket = '<Type your bucket name here>'

###  훈련 잡에 사용할 SageMaker Role 정의 
- 로컬 머신에서 사용할 경우에는 아래의 주석을 해제하시고, SageMaker Execution Role ARN 을 넣어 주세요.

In [67]:
use_local_machine = False
# use_local_machine = True

if use_local_machine:
    role = sagemaker.get_execution_role()
else:
    # role = '<Type Role ARN here>'
    pass

### 로컬 모드 혹은 클라우드 모드에 따른 SageMaker 세션 정의

In [68]:
use_local_mode = False
# use_local_mode = True

if use_local_mode:
    instance_type = "local"
    # instance_type = "local_gpu"
    from sagemaker.local import LocalSession
    sagemaker_session = LocalSession()
    sagemaker_session.config = {'local': {'local_code': True}}
else:
    sagemaker_session = sagemaker.session.Session()
    instance_type = "ml.m5.large"
    



### 하이퍼파라미터 정의

In [70]:
hyperparameters = {
       "scale_pos_weight" : "29",    
        "max_depth": "3",
        "eta": "0.2",
        "objective": "binary:logistic",
        "num_round": "100",
}

### 학습 데이터셋 정의

In [71]:
if use_local_mode:
    local_data_path = "../data/dataset/"
    from pathlib import Path
    file_path = f'file://{Path.cwd()}'
    inputs = file_path.split('1_lab_1_xgboost')[0] + 'data/dataset/'
else:
    data_path=f's3://{bucket}/xgboost/dataset'
    !aws s3 sync ../data/dataset/ $data_path    
    inputs = data_path

print("input data path: ", inputs)                

input data path:  s3://sagemaker-us-east-1-057716757052/xgboost/dataset


### 학습 실행 작업 정의

In [72]:
from sagemaker.xgboost.estimator import XGBoost

estimator = XGBoost(
    entry_point="xgboost_starter_script.py",
    source_dir='src',
    hyperparameters=hyperparameters,
    role=role,
    sagemaker_session=sagemaker_session,
    instance_count= 1,
    instance_type=instance_type,
    framework_version="1.3-1",
)

### 학습 실행

In [73]:
estimator.fit(inputs = {'inputdata': inputs},
                  wait=False)

In [74]:
%%time 

estimator.logs()

2023-05-07 13:18:32 Starting - Starting the training job...
2023-05-07 13:18:54 Starting - Preparing the instances for trainingProfilerReport-1683465510: InProgress
...
2023-05-07 13:19:38 Downloading - Downloading input data......
2023-05-07 13:20:34 Training - Downloading the training image..
2023-05-07 13:21:14 Uploading - Uploading generated training model[2023-05-07 13:20:59.046 ip-10-2-69-169.ec2.internal:7 INFO utils.py:28] RULE_JOB_STOP_SIGNAL_FILENAME: None
[2023-05-07 13:20:59.070 ip-10-2-69-169.ec2.internal:7 INFO profiler_config_parser.py:111] User has disabled profiler.
[2023-05-07:13:20:59:INFO] Imported framework sagemaker_xgboost_container.training
[2023-05-07:13:20:59:INFO] No GPUs detected (normal if no gpus installed)
[2023-05-07:13:20:59:INFO] Invoking user training script.
[2023-05-07:13:20:59:INFO] Installing module with the following command:
/miniconda3/bin/python3 -m pip install . 
Processing /opt/ml/code
  Preparing metadata (setup.py): started
  Preparing met

### 데이터 세트 설명
- 데이터 세트는 블로그 [Architect and build the full machine learning lifecycle with AWS: An end-to-end Amazon SageMaker demo](https://aws.amazon.com/blogs/machine-learning/architect-and-build-the-full-machine-learning-lifecycle-with-amazon-sagemaker/) 에서 사용한 데이터를 사용합니다. 블로그에서는 데이터 세트에 대해서 이렇게 설명 합니다.
- "자동차 보험 청구 사기를 탐지를 위해서 블로그 저자가 데이터를 합성해서 만든 고객과 클레임의 데이터 세트를 사용합니다."

In [78]:
import pandas as pd

In [79]:
train_prep_df = pd.read_csv('../data/dataset/train.csv')
train_prep_df.groupby('fraud').sample(n=5)

fraud  vehicle_claim  total_claim_amount  customer_age  \
518       0   21542.537585        94942.537585            35   
1502      0   17388.642914       115088.642914            69   
2998      0   11163.164243        17463.164243            41   
2813      0   29120.688062        48220.688062            36   
2685      0   18541.583122        43641.583122            45   
1688      1    5481.401810        14381.401810            32   
3946      1    8085.746426        10485.746426            20   
38        1   14423.750237        57223.750237            50   
1778      1   11590.928327        38890.928327            42   
3415      1   15817.129755        26617.129755            49   

      months_as_customer  num_claims_past_year  num_insurers_past_5_years  \
518                  163                     0                          1   
1502                 136                     0                          1   
2998                 175                     0                          1   
2813                 182                     0                          1   
2685                 251                     0                          1   
1688                  66                     0                          1   
3946                  29                     0                          4   
38                    24                     0                          4   
1778                  64                     0                          1   
3415                  95                     0                          1   

      policy_deductable  policy_annual_premium  customer_zip  ...  \
518                 750                   3000         98584  ...   
1502                750                   2850         90063  ...   
2998                750                   3000         86336  ...   
2813                750                   3000         89144  ...   
2685                750                   3000         93711  ...   
1688                750                   3000         89801  ...   
3946                750                   2700         97123  ...   
38                  750                   3000         86404  ...   
1778                750                   3000         92128  ...   
3415                750                   2550         91775  ...   

      collision_type_missing  incident_severity_Major  \
518                        0                        0   
1502                       0                        0   
2998                       0                        0   
2813                       0                        0   
2685                       0                        0   
1688                       0                        0   
3946                       0                        0   
38                         0                        1   
1778                       1                        0   
3415                       1                        0   

      incident_severity_Minor  incident_severity_Totaled  \
518                         0                          1   
1502                        1                          0   
2998                        1                          0   
2813                        0                          1   
2685                        1                          0   
1688                        1                          0   
3946                        1                          0   
38                          0                          0   
1778                        1                          0   
3415                        1                          0   

      authorities_contacted_Ambulance  authorities_contacted_Fire  \
518                                 0                           0   
1502                                0                           0   
2998                                0                           0   
2813                                0                           0   
2685                                0                      

In [80]:
train_prep_df.groupby('fraud').size()

fraud
0    3869
1     131
dtype: int64

#### 데이터 컬럼 설명
- fraud: 보험 청구의 사기 여부 입니다. 1 이면 사기, 0 이면 정상 청구 입니다.
- vehicle_claim: 자동차에 대한 보험 청구액. 값으로서, $1000, $17,638 등이 있습니다.
- total_claim_amount: 전체 보험 청구액 입니다. $21,400, $10,000 등이 있습니다.    
- customer_age: 고객의 나이를 의미 합니다.
- months_as_customer: 고객으로서의 가입 기간을 의미합니다. 단위는 월로서 11, 30, 31 등의 값이 존재 합니다.
- num_claims_past_year: 작년의 보험 청구 수를 의미 합니다. 0, 1, 2, 3, 4, 5, 6 의 값이 존재 합니다.
- num_insurers_past_5_years: 과거 5년 동안의 보험 가입 회사 수를 의미 합니다. 1, 2, 3, 4, 5 의 값이 존재 합니다.
- policy_deductable: 보험의 최소 자기 부담금 입니다. $750, $800 등이 있습니다.    
- policy_annual_premium: 보험의 특약 가입에 대한 금액 입니다. $2000, $3000 등이 있습니다.
- customer_zip: 고객의 집 주소 우편 번호를 의미합니다.
- auto_year: 자동차의 년식을 의미 합니다. 2020, 2019 등이 있습니다.
- num_vehicles_involved: 몇 대의 자동차가 사고에 연관 되었는지 입니다. 1, 2, 3, 4, 5, 6 의 값이 있습니다.
- num_injuries: 몇 명이 상해를 입었는지를 기술합니다. 0, 1, 2, 3, 4, 의 값이 있습니다.
- num_witnesses: 몇 명의 목격자가 있었는지를 기술합니다. 0, 1, 2, 3, 4, 5 의 값이 있습니다.
- injury_claim: 상해에 대한 보험 청구액. \$5,500, \$70,700, \$100,700 등이 있습니다.    
- incident_month: 사고가 발생한 월을 의미합니다. 1~12 값이 존재 합니다.
- incident_day: 사고가 발생한 일자를 의미합니다. 1~31 값이 존재 합니다.
- incident_dow: 사고가 발생한 요일을 의미합니다. 0~6 값이 존재 합니다.
- incident_hour: 사고가 발생한 시간을 의미합니다. 0~23 값이 존재 합니다.
- policy_state: 보험 계약을 한 미국 주(State)를 의미 합니다. CA, WA, AZ, OR, NV, ID 가 존재 합니다.    
- policy_liability: 보험 청구의 한도를 의미 합니다. 예를 들어서 25/50 은  사람 당 상해 한도 $25,000, 사고 당 상해 한도가 $50,000 을  의미합니다. 25/50, 15/30, 30/60, 100/200 의 값이 존재 합니다. 
- customer_gender: 고객의 성별을 의미 합니다. Male, Female, Unkown, Other가 존재 합니다.
- customer_education: 고객의 최종 학력을 의미합니다. Bachelor, High School, Advanced Degree, Associate, Below High School 이 존재 합니다.
- driver_relationship: 보험 계약자와 운전자와의 관계 입니다. Self, Spouse, Child, Other 값이 존재 합니다.
- incident_type: 사고의 종류를 기술합니다. Collision, Break-in, Theft 값이 존재 합니다.
- collision_type: 충돌 타입을 기술합니다. Front, Rear, Side, missing 값이 존재 합니다.
- incident_severity: 사고의 손실 정도 입니다. Minor, Major, Totaled 값이 존재 합니다.
- authorities_contacted: 어떤 관련 기관에 연락을 했는지 입니다. Police, Ambuylance, Fire, None 값이 존재 합니다.
- police_report_available: 경찰 보고서가 존재하는지를 기술합니다. Yes, No 의 값이 있습니다.